src/sync.ts

```diff
@@ -252,6 +275,92 @@ export class SyncApi {
         room.currentState.removeAllListeners(RoomStateEvent.Members);
         room.currentState.removeAllListeners(RoomStateEvent.NewMember);
+        room.currentState.removeAllListeners(RoomStateEvent.Marker);
+    }
+
+    /** When we see the marker state change in the room, we know there is some
+     * new historical messages imported by MSC2716 `/batch_send` somewhere in
+     * the room and we need to throw away the timeline to make sure the
+     * historical messages are shown when we paginate `/messages` again.
+     * @param {Room} room The room where the marker event was sent
+     * @param {MatrixEvent} markerEvent The new marker event
+     * @param {ISetStateOptions} setStateOptions When `timelineWasEmpty` is set
+     * as `true`, the given marker event will be ignored
+    */
+    private onMarkerStateEvent(
+        room: Room,
+        markerEvent: MatrixEvent,
+        { timelineWasEmpty }: IMarkerFoundOptions = {},
+    ): void {
+        // We don't need to refresh the timeline if it was empty before the
+        // marker arrived. This could be happen in a variety of cases:
+        //  1. From the initial sync
+        //  2. If it's from the first state we're seeing after joining the room
+        //  3. Or whether it's coming from `syncFromCache`
+        if (timelineWasEmpty) {
+            logger.debug(
+                `MarkerState: Ignoring markerEventId=${markerEvent.getId()} in roomId=${room.roomId} ` +
+                `because the timeline was empty before the marker arrived which means there is nothing to refresh.`,
+            );
+            return;
+        }
+
+        const isValidMsc2716Event =
+            // Check whether the room version directly supports MSC2716, in
+            // which case, "marker" events are already auth'ed by
+            // power_levels
+            MSC2716_ROOM_VERSIONS.includes(room.getVersion()) ||
+            // MSC2716 is also supported in all existing room versions but
+            // special meaning should only be given to "insertion", "batch",
+            // and "marker" events when they come from the room creator
+            markerEvent.getSender() === room.getRoomCreator();
+
+        if (!isValidMsc2716Event) {
+            logger.debug(
+                `MarkerState: Ignoring markerEventId=${markerEvent.getId()} in roomId=${room.roomId} because ` +
+                `MSC2716 is not supported in the room version or for any room version, the marker wasn't sent ` +
+                `by the room creator.`,
+            );
+        }
+
+        // Don't process a marker event multiple times; we only need to
+        // throw the timeline away once, when we see a marker. All of the
+        // historical content will be in the `/messsages` responses from
+        // here on out.
+        const markerAlreadyProcessed = markerEvent.getId() === room.getLastMarkerEventIdProcessed();
+        if (markerAlreadyProcessed) {
```

Comment:
We moved to marker events as state with each one having a unique `state_key` so they all appear readily to homeservers/clients without getting lost in timeline gaps. This means there will be multiple marker events in the state at a time.

TODO: Add a test to make sure this plays well with multiple markers and we don't just get in a loop of processing them back and fourth as they invalidate each other. In other words, does `onMarkerStateEvent` only get called for new marker state, or all marker events from any time we see it (like when we jump to a permalink with `/context` for example)?

I think `setStateEvents` (which emits `RoomStateEvent.Marker` -> `onMarkerStateEvent`) currently doesn't discriminate between new vs old state we already have based on the comment doc, "Add an array of one or more state MatrixEvents, overwriting any existing state with the same {type, stateKey} tuple." So we do have to worry about this.

Ideally, we could adjust `setStateEvents` to store the state before applying the new state, then we can compare and only emit `RoomStateEvent.Marker` if we didn't already have it. This way, we can also get rid of the `getLastMarkerEventIdProcessed` stuff.

Otherwise, we have to do keep track of things we already processed differently which doesn't sound fun. List of every marker event, using `origin_server_ts` which is flawed, etc.

Reply:
> I think `setStateEvents` (which emits `RoomStateEvent.Marker` -> `onMarkerStateEvent`) currently doesn't discriminate between new vs old state

We do discriminate when `timelineWasEmpty` and ignore those which seems to do exactly what we need.

In all of my tests, we don't ever trigger this scenario and now looking at it, I'm not sure how we would. I'm going to remove for now. If we ever need it again, we'll understand why and can write a test for it ⏩